tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp

```diff
@@ -78,11 +86,38 @@ void kernel_main() {
     noc_semaphore_set(in0_mcast_receiver_semaphore_addr_ptr, VALID);
 
     cb_reserve_back(cb_id_in2, batch * in0_block_num_tiles);
-    uint32_t local_read_addr = get_read_ptr(cb_id_in2);
+
+    uint32_t local_read_addr;
+    if constexpr (in4_block_num_tiles > 0) {
+        cb_reserve_back(cb_id_in4, batch * in4_block_num_tiles);
+        local_read_addr = get_read_ptr(cb_id_in4);
+    } else {
+        local_read_addr = get_read_ptr(cb_id_in2);
+    }
 
     for (uint32_t b = 0; b < batch; ++b) {
-        for(uint32_t block = 0; block < num_blocks; ++block) {
-            if (block == sender_id) {
+        for (uint32_t block = 0; block < num_blocks; ++block) {
+            const uint32_t block_id = block / num_blocks_per_shard;
+            if (block_id == sender_id) {
+                if constexpr (in4_block_num_tiles > 0) {
+                    // In case shard_height_in_tiles is > 32 and in0_block_w != shard_width_in_tiles
+                    // We need to to copy the in0_block_w columns of shard data into a temporary cb
+                    // and then mcast from there
+                    uint32_t shard_read_addr = get_read_ptr(cb_id_in2);
+                    uint32_t write_ptr_cb4 = get_write_ptr(cb_id_in4);
```

Comment:
`get_write_ptr` should follow `cb_reserve_back` in general. Unless you know that pushing won't affect your ptrs.

Reply:
thx